File: ios/sensors/HrStreamManager.swift
import Foundation
import PolarBleSdk
import RxSwift
import React

class HrStreamManager {

    private weak var bridge: PolarBridge?
    private var api: PolarBleApi?

    private var hrDisposable: Disposable?
    private var isHrStreaming = false

    private var hrBuffer: [[String: Any]] = []
    private let hrBufferQueue = DispatchQueue(label: "com.polarbridge.hrBufferQueue")
    private var hrFlushTimer: Timer?

    private let SENSOR_BUFFER_MS: TimeInterval = 10_000
    private let disposeBag = DisposeBag()

    init(api: PolarBleApi?, bridge: PolarBridge) {
        self.api = api
        self.bridge = bridge
    }

    // MARK: - Public API

    func fetchHrData(_ deviceId: String, bufferMs: NSNumber?) {
        NSLog("HrStreamManager: Fetch HR Data called on: \(deviceId) bufferMs: \(String(describing: bufferMs))")

        let resolvedBufferMs: TimeInterval
        if let number = bufferMs, !(number is NSNull) {
            resolvedBufferMs = number.doubleValue >= 0 ? number.doubleValue : SENSOR_BUFFER_MS
        } else {
            resolvedBufferMs = SENSOR_BUFFER_MS
        }

        guard let api = api else {
            NSLog("HrStreamManager: Polar API not initialized")
            return
        }

        // Stop if already running
        if isHrStreaming {
            disposeHrStream()
            stopHrFlushTimer()
            flushHrBuffer()
            isHrStreaming = false

            bridge?.sendEvent(
                withName: PolarEvent.PolarHrComplete.rawValue,
                body: ["message": "HR Stream stopped"]
            )
            return
        }

        isHrStreaming = true
        startHrFlushTimer(bufferMs: resolvedBufferMs)

        hrDisposable = api.startHrStreaming(deviceId)
            .observe(on: MainScheduler.instance)
            .subscribe(
                onNext: { [weak self] hrData in
                    guard let self = self else { return }

                    self.hrBufferQueue.async {
                        for sample in hrData {
                            var event: [String: Any] = [:]
                            event["hr"] = sample.hr
                            event["rrsMs"] = sample.rrsMs
                            event["rrAvailable"] = sample.rrAvailable
                            event["contactStatus"] = sample.contactStatus
                            event["contactStatusSupported"] = sample.contactStatusSupported
                            event["timestamp"] = Date().timeIntervalSince1970 * 1000

                            self.hrBuffer.append(event)
                        }
                    }
                },
                onError: { [weak self] error in
                    guard let self = self else { return }

                    self.stopHrFlushTimer()
                    self.flushHrBuffer()
                    self.isHrStreaming = false

                    self.bridge?.sendEvent(
                        withName: PolarEvent.PolarHrError.rawValue,
                        body: ["error": error.localizedDescription]
                    )
                },
                onCompleted: { [weak self] in
                    guard let self = self else { return }

                    self.stopHrFlushTimer()
                    self.flushHrBuffer()
                    self.isHrStreaming = false

                    self.bridge?.sendEvent(
                        withName: PolarEvent.PolarHrComplete.rawValue,
                        body: ["message": "HR stream complete"]
                    )
                }
            )

        hrDisposable?.disposed(by: disposeBag)
    }

    func disposeHrStream() {
        isHrStreaming = false
        hrDisposable?.dispose()
    }

    // MARK: - Buffer

    private func startHrFlushTimer(bufferMs: TimeInterval) {
        stopHrFlushTimer()

        let intervalSeconds = bufferMs / 1000.0

        DispatchQueue.main.async {
            self.hrFlushTimer = Timer(
                timeInterval: intervalSeconds,
                repeats: true
            ) { [weak self] _ in
                self?.flushHrBuffer()
            }

            RunLoop.main.add(self.hrFlushTimer!, forMode: .common)
        }
    }

    private func stopHrFlushTimer() {
        DispatchQueue.main.async {
            self.hrFlushTimer?.invalidate()
            self.hrFlushTimer = nil
        }
    }

    private func flushHrBuffer() {
        hrBufferQueue.async {
            guard !self.hrBuffer.isEmpty else { return }

            let events = self.hrBuffer
            self.hrBuffer.removeAll()

            DispatchQueue.main.async {
                for event in events {
                    self.bridge?.sendEvent(
                        withName: PolarEvent.PolarHrData.rawValue,
                        body: event
                    )
                }
            }
        }
    }
}

File: android/src/main/java/com/polarbridge/sensors/HrStreamManager.kt
package com.polarbridge.sensors

import android.util.Log
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import com.polar.sdk.api.PolarBleApi
import com.polar.sdk.api.errors.PolarInvalidArgument
import com.polar.sdk.api.model.PolarHrData
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import java.util.concurrent.TimeUnit

class HrStreamManager(
  private val api: PolarBleApi,
  private val sendEvent: (String, WritableMap) -> Unit,
  private val TAG: String = "HrStreamManager"
) {
  private var hrDisposable: Disposable? = null

  fun fetchHrData(deviceId: String, ms: Double?, sensorBufferMs: Long) {
    Log.d(TAG, "Fetch Heart Data called on: $deviceId ")

    val bufferMs = ms?.toLong()?.takeIf { it >= 0 } ?: sensorBufferMs
    val isDisposed = hrDisposable?.isDisposed ?: true

    try {
      if (isDisposed) {
        hrDisposable = api.startHrStreaming(deviceId)
          .flatMapIterable { hrData ->
            hrData.samples.map { sample ->
              HrSampleWithTimestamp(
                data = sample,
                timestampMs = System.currentTimeMillis()
              )
            }
          }
          .buffer(bufferMs, TimeUnit.MILLISECONDS)
          .observeOn(AndroidSchedulers.mainThread())
          .subscribe(
            { samples ->
              if (samples.isEmpty()) return@subscribe

              Log.d(TAG, "Flushing HR buffer (${samples.size} samples)")

              samples.forEach { item ->
                val sample = item.data

                val event = Arguments.createMap().apply {
                  putInt("hr", sample.hr)

                  val rrsArray = Arguments.createArray()
                  sample.rrsMs.forEach { rrsArray.pushInt(it) }

                  putArray("rrsMs", rrsArray)
                  putBoolean("rrAvailable", sample.rrAvailable)
                  putBoolean("contactStatus", sample.contactStatus)
                  putBoolean("contactStatusSupported", sample.contactStatusSupported)
                  putDouble("timestamp", item.timestampMs.toDouble())
                }

                sendEvent("PolarHrData", event)
              }
            },
            { error ->
              Log.e(TAG, "HR stream failed. Reason $error")

              val errorEvent = Arguments.createMap().apply {
                putString("error", error.toString())
              }

              sendEvent("PolarHrError", errorEvent)
            },
            {
              Log.d(TAG, "HR stream complete")

              val completeEvent = Arguments.createMap().apply {
                putString("message", "HR stream complete")
              }

              sendEvent("PolarHrComplete", completeEvent)
            }
          )
      } else {
        disposeHrStream()
        Log.d(TAG, "HR stream stopped")
      }
    } catch (e: PolarInvalidArgument) {
      Log.e(TAG, "Failed to fetch HR Data. Reason $e")
    }
  }

  fun disposeHrStream() {
    try {
      hrDisposable?.dispose()
    } catch (e: Exception) {
      // Bluetooth service is dead, cleanup command failed.
      Log.w(TAG, "Failed to dispose HR stream cleanly. Bluetooth service likely died.", e)
    } finally {
      // Nullify the reference to flag that the stream is gone,
      // preventing memory leaks or further attempts to dispose of a dead stream.
      hrDisposable = null
    }
  }

  data class HrSampleWithTimestamp(
    val data: PolarHrData.PolarHrSample,
    val timestampMs: Long
  )
}
